## Part 2
**TASK 6:Customer Segmentation and Value-Based Structuring**

In [34]:
# Step 1: Import pandas for Data analysis and manipulation
import pandas as pd


In [35]:
URL = "https://raw.githubusercontent.com/YBI-Foundation/Dataset/main/Bank%20Churn%20Modelling.csv"

# Load the dataset into a DataFrame
bank_df = pd.read_csv( URL)
# Step 3: Display the first five rows
print("First 5 rows of the dataset:")
display(bank_df.head())

df_extra = pd.read_csv("Bank_Churn_Modelling_extra_UPDATED_v2 (1).csv")
df_extra.head()


First 5 rows of the dataset:


,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,Num Of Products,Has Credit Card,Is Active Member,Estimated Salary,Churn
0,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,Num Of Products,Has Credit Card,Is Active Member,Estimated Salary,Churn
0,15647819,Dlamini,536.0,Germany,Male,72.0,9,116534.69,4,1,1,163135.72,0
1,15587369,Jones,572.0,France,Female,NaN,4,153720.31,3,0,0,156006.35,0
2,15661951,VanNiekerk,357.0,Germany,Female,85.0,5,66304.03,2,0,0,166888.24,0
3,15567632,Jones,769.0,Spain,Male,NaN,2,33726.41,4,0,0,191141.68,1
4,15636646,Dlamini,378.0,Germany,NaN,62.0,9,0.00,1,1,0,197821.39,1


In [36]:
bank_df = pd.read_csv("https://raw.githubusercontent.com/YBI-Foundation/Dataset/main/Bank%20Churn%20Modelling.csv")
df_extra = pd.read_csv("Bank_Churn_Modelling_extra_UPDATED_v2 (1).csv")

In [39]:
combined_df = pd.concat([bank_df, df_extra], ignore_index=True)
combined_df.head()

,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,Num Of Products,Has Credit Card,Is Active Member,Estimated Salary,Churn
0,15634602,Hargrave,619.0,France,Female,42.0,2,0.00,1,1,1,101348.88,1
1,15647311,Hill,608.0,Spain,Female,41.0,1,83807.86,1,0,1,112542.58,0
2,15619304,Onio,502.0,France,Female,42.0,8,159660.80,3,1,0,113931.57,1
3,15701354,Boni,699.0,France,Female,39.0,1,0.00,2,0,0,93826.63,0
4,15737888,Mitchell,850.0,Spain,Female,43.0,2,125510.82,1,1,1,79084.10,0


In [41]:
# Fill numerical columns with mean
combined_df['CreditScore'] = combined_df['CreditScore'].fillna(combined_df['CreditScore'].mean())
combined_df['Age'] = combined_df['Age'].fillna(combined_df['Age'].mean())
combined_df['Balance'] = combined_df['Balance'].fillna(combined_df['Balance'].mean())
combined_df['Estimated Salary'] = combined_df['Estimated Salary'].fillna(combined_df['Estimated Salary'].mean())

# Fill categorical columns with mode
combined_df['Geography'] = combined_df['Geography'].fillna(combined_df['Geography'].mode()[0])
combined_df['Gender'] = combined_df['Gender'].fillna(combined_df['Gender'].mode()[0])

# Verify
print(combined_df.isnull().sum())

CustomerId          0
Surname             0
CreditScore         0
Geography           0
Gender              0
Age                 0
Tenure              0
Balance             0
Num Of Products     0
Has Credit Card     0
Is Active Member    0
Estimated Salary    0
Churn               0
dtype: int64


**TASK 6.1:Customer Segmentation (Age-Based)**

In [45]:
df = clean_df

In [49]:
df['Age_Group'] = pd.cut(
    df['Age'],
    bins=[0, 30, 50, 100],
    labels=['Young (18–30)', 'Middle-aged (31–50)', 'Older (51+)']
)

In [47]:
df[['Age', 'Age_Group']].head()

,Age,Age_Group
0,42.0,Middle-aged
1,41.0,Middle-aged
2,42.0,Middle-aged
3,39.0,Middle-aged
4,43.0,Middle-aged


In [55]:
# Count customers in each age group
df['Age_Group'].value_counts()

Age_Group
Middle-aged (31–50)    6784
Young (18–30)          1973
Older (51+)            1281
Name: count, dtype: int64

Customers were segmented into age groups to better understand demographic differences.

The distribution of customers across age groups is as follows:

- Middle-aged (31–50): 6,784 customers  
- Young (18–30): 1,973 customers  
- Older (51+): 1,281 customers  

The average churn rate was calculated for each age group. This helps identify which age group is more likely to leave the bank.

**TASK 6.2: Value-Based Segmentation**

#### Value-Based Segmentation Strategy
Customers were segmented based on their **account balance**, which reflects their financial value to the bank.


In [56]:
# Create value-based segments using Balance
df['Value_Segment'] = pd.cut(
    df['Balance'],
    bins=[-1, 50000, 150000, df['Balance'].max()],
    labels=['Low Value', 'Medium Value', 'High Value']
)

Customers were categorised based on their **account balance**, which reflects their financial value to the bank.

### Segments Defined

- Low Value  
- Medium Value  
- High Value  

These segments allow the bank to identify customers with different financial importance.

In [57]:
df[['Balance', 'Value_Segment']].head()

,Balance,Value_Segment
0,0.00,Low Value
1,83807.86,Medium Value
2,159660.80,High Value
3,0.00,Low Value
4,125510.82,Medium Value


The number of customers in each value segment was calculated to understand how customers are distributed based on their financial value.

In [58]:
df['Value_Segment'].value_counts()

Value_Segment
Medium Value    5353
Low Value       3701
High Value       984
Name: count, dtype: int64

The majority of customers fall within the medium value segment, indicating that most customers maintain moderate account balances.  
A significant portion of customers are in the low value segment, suggesting lower financial engagement.  
The high value segment is the smallest, representing customers with the highest financial value to the bank.

**TASK 7: Group Aggregation Analysis**

In [63]:
df.groupby('Age_Group', observed=True)['Balance'].mean()

Age_Group
Young (18–30)          73416.559584
Middle-aged (31–50)    76886.081224
Older (51+)            80323.412803
Name: Balance, dtype: float64

 Older customers (51+) have the highest average balance (~80,323), indicating higher financial value.  


In [64]:
df.groupby('Churn')['CreditScore'].mean()

Churn
0    651.622094
1    644.548905
Name: CreditScore, dtype: float64

Customers who remained (Churn = 0) have a slightly higher average credit score. 
This indicates that lower credit scores may be associated with a higher likelihood of churn.


In [65]:
df.groupby(['Geography', 'Gender'], observed=True)['Churn'].mean()

Geography  Gender
France     Female    0.203622
           Male      0.127721
Germany    Female    0.376872
           Male      0.279489
Spain      Female    0.212980
           Male      0.132279
Name: Churn, dtype: float64

Germany has the highest churn rates for both females.  
France and Spain show lower churn rates overall.

In [66]:
df.groupby('Value_Segment', observed=True).agg({
    'Balance': 'mean',
    'Estimated Salary': 'mean',
    'CustomerId': 'count',
    'Churn': 'mean'
})

,Balance,Estimated Salary,CustomerId,Churn
Value_Segment,,,,
Low Value,856.630322,99035.964335,3701,0.143475
Medium Value,112374.474600,100840.329143,5353,0.241734
High Value,167306.288140,99859.312632,984,0.233740


In [67]:
df.groupby(['Age_Group', 'Value_Segment'], observed=True).agg({
    'Balance': 'mean',
    'CustomerId': 'count',
    'Churn': 'mean'
})

Balance  CustomerId     Churn
Age_Group           Value_Segment                                     
Young (18–30)       Low Value         716.281339         784  0.042092
                    Medium Value   112149.589859         994  0.091549
                    High Value     168269.821385         195  0.128205
Middle-aged (31–50) Low Value         821.660185        2486  0.139179
                    Medium Value   112474.212264        3632  0.232654
                    High Value     166736.019309         666  0.214715
Older (51+)         Low Value        1313.635800         431  0.352668
                    Medium Value   112183.674376         727  0.492435
                    High Value     168866.532520         123  0.504065

**Age is a strong driver of churn**, with older customers more likely to leave.
**High-value older customers represent the highest-risk segment**, combining high financial value with high churn probability.
Medium-value customers consistently show relatively high churn, indicating a potentially unstable group.


**TASK 8: Pivot Table Construction**

In [68]:
pivot_table = df.pivot_table(
    values='Churn',
    index='Geography',
    columns='Gender',
    aggfunc='mean'
)

pivot_table

Gender,Female,Male
Geography,,
France,0.203622,0.127721
Germany,0.376872,0.279489
Spain,0.212980,0.132279


A pivot table was created to analyse churn behaviour across two categorical variables:
Geography and Gender

In [69]:
pivot_table.round(3)

Gender,Female,Male
Geography,,
France,0.204,0.128
Germany,0.377,0.279
Spain,0.213,0.132


The pivot table shows churn rates across different regions and genders.
Differences in churn rates indicate how customer behaviour varies across demographic groups.
Higher values suggest higher likelihood of customers leaving.



Germany shows the highest churn rates for both females (0.377) and males (0.279), indicating a higher risk of customer loss in this region.  
France and Spain have lower churn rates overall, suggesting more stable customer retention.  

#### Interpretation
The bank should prioritise customer retention efforts in **Germany**, especially focusing on **female customers**.  
Tailored strategies based on demographic segments are likely to be more effective than general approaches.
